src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/InlineCompletion/InlineCompletionEndPoint.cs

```diff
@@ -0,0 +1,240 @@
+        var list = await response.Returning<InlineCompletionList>(cancellationToken);
+        if (list == null)
+        {
+            _logger.LogInformation($"Did not get any inline completions from delegation.");
+            return null;
+        }
+
+        var items = new List<InlineCompletionItem>();
+        var csharpDocOptions = codeDocument.GetCSharpDocument();
+        foreach (var item in list.Items)
+        {
+            var containsSnippet = item.TextFormat == InsertTextFormat.Snippet;
+            var range = item.Range ?? new Range { Start = projectedPosition, End = projectedPosition };
+
+            if (!_documentMappingService.TryMapFromProjectedDocumentRange(codeDocument, range, out var rangeInRazorDoc))
+            {
+                _logger.LogWarning($"Could not remap projected range {range} to razor document");
+                continue;
+            }
+
+            using var formattingContext = FormattingContext.Create(request.TextDocument.Uri, document, codeDocument, s_defaultFormattingOptions, _adhocWorkspaceFactory, isFormatOnType: true, automaticallyAddUsings: false);
+            if (!TryGetSnippetWithAdjustedIndentation(formattingContext, item.Text, hostDocumentIndex, out var newSnippetText))
+            {
+                continue;
+            }
+
+            var remappedItem = new InlineCompletionItem
+            {
+                Command = item.Command,
+                Range = rangeInRazorDoc,
+                Text = newSnippetText.ToString(),
+                TextFormat = item.TextFormat,
+            };
+            items.Add(remappedItem);
+        }
+
+        _logger.LogInformation($"Returning items.");
+        return new InlineCompletionList
+        {
+            Items = items.ToArray()
+        };
+    }
+
+    private static bool TryGetSnippetWithAdjustedIndentation(FormattingContext formattingContext, string snippetText, int hostDocumentIndex, [NotNullWhen(true)] out string? newSnippetText)
+    {
+        newSnippetText = null;
+        if (!formattingContext.TryGetFormattingSpan(hostDocumentIndex, out var formattingSpan))
+        {
+            return false;
+        }
+
+        // Take the amount of indentation razor and html are adding, then remove the amount of C# indentation that is 'hidden'.
+        // This should give us the desired base indentation that must be applied to each line.
+        var razorAndHtmlContributionsToIndentation = formattingSpan.RazorIndentationLevel + formattingSpan.HtmlIndentationLevel;
+        var amountToAddToCSharpIndentation = razorAndHtmlContributionsToIndentation - formattingSpan.MinCSharpIndentLevel;
+
+        var snippetSourceText = SourceText.From(snippetText);
+        List<TextChange> indentationChanges = new();
+        // Adjust each line, skipping the first since it must start at the snippet keyword.
+        foreach (var line in snippetSourceText.Lines.Skip(1))
```

Comment:
Hmm do we have to skip the first line?

Asking because I'd hope that whatever path we take in formatting a snippet could be pulled into the formatting engine so someone can do something like `razorFormatter.FormatSnippet(...)`. Only bringing it up because I can envisions us wanting to re-use this logic (override completion comes to mind but of course that only has $0 typically).

Reply:
It's possible, but would need to be special cased.  Because the text edit that C# returns starts at the snippet keyword, and only replaces the snippet keyword.  So we'd have to extend the range to the rest of the line (including possible text before the keyword), then modify that indentation, which seemed more complicated than it was worth.  Especially since the range for inlien completions should start at that keyword
https://github.com/microsoft/vscode/blob/075ba020e8493f40dba89891b1a08453f2c067e9/src/vscode-dts/vscode.proposed.inlineCompletions.d.ts#L82

Reply:
> I think combining the logic for formatting snippets (at least for C#) is a good idea - however I would like to battle test this formatting code before we unify things

Totally fair!